### ISSUE TITLE
Add feature toggle to disable autoImports

### ISSUE THREAD

**Issue by alSergey** [nuxt/nuxt#26525]
### Describe the feature

Hello, thank you very much for the product you make. 

When working with `nuxt`, I don't really like the concept of auto-imports and would like to disable it. I have several reasons why I want to do this:
1) Auto imports are secret magic. It’s not entirely clear what you are importing and from where
2) I use `jest` in my project and there are big problems with it when setting up mocks for `#imports`

I saw a previous [issue](https://github.com/nuxt/nuxt/issues/14505) on this topic, but as I see it did not end in anything. Currently, there are settings to disable auto-imports of components, but there are no settings to disable other auto-imports: `vue` re-export, `@unhead/vue` re-export, `composables` folder re-export in a project, `utils` folder re-export in a project.

I would like to have some setting that allows me to disable all re-exports from `#imports` except `nuxt` re-exports.

### Additional information

- [ ] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

**Comment by alSergey** [comment]
I found a solution to remove `composables`, `utils` and `vue` re-exports, but it doesn't seem like a good solution. However, I still haven’t found how to remove `@unhead/vue` re-exports.

```ts

export default defineNuxtConfig({
		hooks: {
			'imports:dirs': (dirs) => {
				dirs.splice(0, dirs.lenght)
			},
			'imports:sources': (preset) => {
				preset.splice(0, preset.length, ...preset.filter(({ from }) => from !== 'vue'))
			},
		},
})
```

**Comment by sriram-500apps** [comment]
@alSergey If you want to disable auto-imports of composables and utils you can disable in `nuxt.config.ts`

```
export default defineNuxtConfig({
  imports: {
    autoImport: false
  }
})
```
check for more - https://nuxt.com/docs/guide/concepts/auto-imports#disabling-auto-imports

disable auto-import of components - https://nuxt.com/docs/guide/concepts/auto-imports#auto-imported-components

**Comment by pothuraju-dev** [comment]
@alSergey If you want to disable auto-imports of composables and utils you can disable in nuxt.config.ts

export default defineNuxtConfig({
  imports: {
    autoImport: false
  }
})
check for more - https://nuxt.com/docs/guide/concepts/auto-imports#disabling-auto-imports

disable auto-import of components - https://nuxt.com/docs/guide/concepts/auto-imports#auto-imported-components

**Comment by alSergey** [comment]
@pothuraju-dev Hi, setting autoImports does not disable importing `composables` and `utils` folders from the project. You can check it out in this [repository](https://github.com/alSergey/nuxt-autoimports-repro). I looked [at your code](https://github.com/nuxt/nuxt/blob/ae72c4a60d0ca7cd4fe2f51439d1001d8d79de9e/packages/nuxt/src/imports/module.ts#L54) and found that you do not check for this flag, but simply always add auto-imports for these folders.

And is there any option to disable `vue` and `@unhead/vue` auto-imports?

**Comment by alSergey** [comment]
I get it, I used the wrong terminology. When I talked about auto imports, I meant re-exports from `#imports`. I have already disabled auto-imports on the project, now I want to disable unnecessary re-exports.

**That is, now my `imports.d.ts` file looks like this.**

```ts
export { isVue2, isVue3 } from 'vue-demi';
export { defineNuxtLink } from '#app/components/nuxt-link';
export { useNuxtApp, tryUseNuxtApp, defineNuxtPlugin, definePayloadPlugin, useRuntimeConfig, defineAppConfig } from '#app/nuxt';
export { requestIdleCallback, cancelIdleCallback } from '#app/compat/idle-callback';
export { setInterval } from '#app/compat/interval';
export { useAppConfig, updateAppConfig } from '#app/config';
export { defineNuxtComponent } from '#app/composables/component';
export { useAsyncData, useLazyAsyncData, useNuxtData, refreshNuxtData, clearNuxtData } from '#app/composables/asyncData';
export { useHydration } from '#app/composables/hydrate';
export { callOnce } from '#app/composables/once';
export { useState, clearNuxtState } from '#app/composables/state';
export { clearError, createError, isNuxtError, showError, useError } from '#app/composables/error';
export { useFetch, useLazyFetch } from '#app/composables/fetch';
export { useCookie, refreshCookie } from '#app/composables/cookie';
export { prerenderRoutes, useRequestHeader, useRequestHeaders, useRequestEvent, useRequestFetch, setResponseStatus } from '#app/composables/ssr';
export { onNuxtReady } from '#app/composables/ready';
export { preloadComponents, prefetchComponents, preloadRouteComponents } from '#app/composables/preload';
export { abortNavigation, addRouteMiddleware, defineNuxtRouteMiddleware, setPageLayout, navigateTo, useRoute, useRouter, onBeforeRouteLeave, onBeforeRouteUpdate } from '#app/composables/router';
export { isPrerendered, loadPayload, preloadPayload, definePayloadReducer, definePayloadReviver } from '#app/composables/payload';
export { useLoadingIndicator } from '#app/composables/loading-indicator';
export { getAppManifest, getRouteRules } from '#app/composables/manifest';
export { reloadNuxtApp } from '#app/composables/chunk';
export { useRequestURL } from '#app/composables/url';
export { usePreviewMode } from '#app/composables/preview';
export { useId } from '#app/composables/id';
export { withCtx, withDirectives, withKeys, withMemo, withModifiers, withScopeId, onActivated, onBeforeMount, onBeforeUnmount, onBeforeUpdate, onDeactivated, onErrorCaptured, onMounted, onRenderTracked, onRenderTriggered, onServerPrefetch, onUnmounted, onUpdated, computed, customRef, isProxy, isReactive, isReadonly, isRef, markRaw, proxyRefs, reactive, readonly, ref, shallowReactive, shallowReadonly, shallowRef, toRaw, toRef, toRefs, triggerRef, unref, watch, watchEffect, watchPostEffect, watchSyncEffect, isShallow, effect, effectScope, getCurrentScope, onScopeDispose, defineComponent, defineAsyncComponent, resolveComponent, getCurrentInstance, h, inject, hasInjectionContext, nextTick, provide, defineModel, defineOptions, defineSlots, mergeModels, toValue, useModel, useAttrs, useCssModule, useCssVars, useSlots, useTransitionState, Component, ComponentPublicInstance, ComputedRef, ExtractPropTypes, ExtractPublicPropTypes, InjectionKey, PropType, Ref, MaybeRef, MaybeRefOrGetter, VNode } from 'vue';
export { injectHead, useHead, useSeoMeta, useHeadSafe, useServerHead, useServerSeoMeta, useServerHeadSafe } from '@unhead/vue';
export { useStepLoading } from '../composables/use-step-loading';
```

**And I want the imports.d.ts file to look like this.**

```ts
export { isVue2, isVue3 } from 'vue-demi';
export { defineNuxtLink } from '#app/components/nuxt-link';
export { useNuxtApp, tryUseNuxtApp, defineNuxtPlugin, definePayloadPlugin, useRuntimeConfig, defineAppConfig } from '#app/nuxt';
export { requestIdleCallback, cancelIdleCallback } from '#app/compat/idle-callback';
export { setInterval } from '#app/compat/interval';
export { useAppConfig, updateAppConfig } from '#app/config';
export { defineNuxtComponent } from '#app/composables/component';
export { useAsyncData, useLazyAsyncData, useNuxtData, refreshNuxtData, clearNuxtData } from '#app/composables/asyncData';
export { useHydration } from '#app/composables/hydrate';
export { callOnce } from '#app/composables/once';
export { useState, clearNuxtState } from '#app/composables/state';
export { clearError, createError, isNuxtError, showError, useError } from '#app/composables/error';
export { useFetch, useLazyFetch } from '#app/composables/fetch';
export { useCookie, refreshCookie } from '#app/composables/cookie';
export { prerenderRoutes, useRequestHeader, useRequestHeaders, useRequestEvent, useRequestFetch, setResponseStatus } from '#app/composables/ssr';
export { onNuxtReady } from '#app/composables/ready';
export { preloadComponents, prefetchComponents, preloadRouteComponents } from '#app/composables/preload';
export { abortNavigation, addRouteMiddleware, defineNuxtRouteMiddleware, setPageLayout, navigateTo, useRoute, useRouter, onBeforeRouteLeave, onBeforeRouteUpdate } from '#app/composables/router';
export { isPrerendered, loadPayload, preloadPayload, definePayloadReducer, definePayloadReviver } from '#app/composables/payload';
export { useLoadingIndicator } from '#app/composables/loading-indicator';
export { getAppManifest, getRouteRules } from '#app/composables/manifest';
export { reloadNuxtApp } from '#app/composables/chunk';
export { useRequestURL } from '#app/composables/url';
export { usePreviewMode } from '#app/composables/preview';
export { useId } from '#app/composables/id';
```

That is, only `nuxt` re-exports remain in this file.

**Comment by pothuraju-dev** [comment]
I think there is no option to avoid auto-importing of `@unhead/vue` from `#imports`. I have observed that it has been hardcoded in nuxt. you can check below code

https://github.com/nuxt/nuxt/blob/ae72c4a60d0ca7cd4fe2f51439d1001d8d79de9e/packages/nuxt/src/head/module.ts#L38-L50

Added to `nuxt/core` directly
https://github.com/nuxt/nuxt/blob/ae72c4a60d0ca7cd4fe2f51439d1001d8d79de9e/packages/nuxt/src/core/nuxt.ts#L16
https://github.com/nuxt/nuxt/blob/ae72c4a60d0ca7cd4fe2f51439d1001d8d79de9e/packages/nuxt/src/core/nuxt.ts#L545

**Comment by alSergey** [comment]
Yes, I also didn’t find an option to disable re-exports. That is why I suggested the option of adding the toggle feature for this.

**Comment by danielroe** [comment]
This is absolutely possible to do. I think it makes sense document how to do it (I can come back later with a config example). If this is requested by enough people we can consider a toggle but I'm not convinced that would be the right thing to do.

cc: @antfu

**Comment by alSergey** [comment]
@danielroe hi, I would be very grateful if you could send me an example config on how to do this. Thanks in advance 😊

**Comment by antfu** [comment]
I am curious about what's the problem with re-exporting? It's a type DTS file, and it's not global; supposedly, it doesn't affect how things work, isn't it?

**Comment by alSergey** [comment]
Yes, this file really does not affect the application itself, but it does affect the writing of tests (I wrote [about this](https://github.com/nuxt/nuxt/issues/26525#issue-2211938979) at the very beginning). I may not to use imports from `#imports`, but there are a few reasons why I want to disable this feature:
1) I have several projects on which not only I work. Not everyone is careful to use `#imports`
2) The IDE suggests importing from `#imports` rather than from `vue` or `@unhead/vue`. I can try to reconfigure this, but then we return to the first point

It is for these reasons that I want to disable everything from `#imports` except `nuxt`

**Comment by antoinerey** [comment]
For what it's worth, here are my two cents. 

As a bit of context, I've been using Nuxt professionally for six years now. I've witnessed the introduction of the opt-in auto-import feature in Nuxt 2, and then the mandatory auto-import feature in Nuxt 3.

When Nuxt 3 was officially released, about a year ago, I started to take part in our project migration to modernize what was running on Nuxt 2 at that time. To give some idea of the scale, about 30 front-end engineers work full-time on that project, building customer-facing pages in dozens of different scopes.

With the background being put, let's jump on the actual topic. My main struggle with the auto-import feature is the fact that _I feel like_ we're going back in time to the global variables era. Having to manually ensure that there is no conflict between two exported entities feels wrong to me now that ESM has officially standardized the import/export mechanism and takes care of managing those conflicts for us.

That said, my feelings on this auto-import feature evolved a lot during the past few months. I have to admit that I wrongly pushed back the idea without giving it its proper chance when it was introduced in Nuxt 2 and then kept my strong opinion with Nuxt 3. But, I recently started a _small_ project with Nuxt 3, and have been surprised that when there are no chances of conflicts between two exports, the developer experience is pretty delightful. 

However, I'm still convinced that it can't work on large projects with lots of people contributing on a daily basis on dozens of different scopes. On our side, we made the choice to modularize all our application scopes so that our developer teams can be autonomous in their implementations. For example, the "payment" team owns a module dedicated to payment features, and the "authentication" team owns a module dedicated to login and register. 

By using the auto-import feature, both teams would have to make sure not to expose components, composables, utils, or whatever, named the same way or one would override the other. And this increases exponentially the more teams work in parallel. Of course, there is always the ability to prefix the exported members with the scope they belong to, but again it _feels_ wrong to me.

All-in-all, I do think that the auto-import feature is valuable and enhances the developer experience in small-to-medium projects. But it tends to go the other way in large projects, making the developer experience worse because of the workarounds we have to put in place to prevent conflicts.

So, I would also be in favour of turning the auto-import feature into an opt-out mechanism so we could turn it off when we feel like it.

**Comment by antfu** [comment]
I totally understand the concerns - we **do** let you disable auto-import if needed.

Exports in `#imports` to me are not considered as `auto-import`, but more like "explicit auto resolve" which does not really inference the "globals".

We can't just disable `#imports` as modules are relying on it to resolve the imports they need. And resolved paths are not always as simple as `import { useHead } from '@unhead/vue'` but could be something like `import { useHead } from '../node_modules/module-foo/node_modules/@unhead/vue'`, which are non-hoisting friendly that hard to write manually (also not always deterministic)

If you don't want your IDE to suggest that, I am sure there would be ways to configure it. For example, in VS Code you could do this with `settings.json`:

```json
  "typescript.preferences.autoImportFileExcludePatterns": [
    "#imports",
  ],
```

I hope that would resolve the issue you have.

**Comment by alSergey** [comment]
@antfu hi, as I [already wrote above](https://github.com/nuxt/nuxt/issues/26525#issuecomment-2027164560), the option of setting up an IDE is not suitable for me. I have several projects that are worked on by different people with different IDEs and there is always a chance that someone accidentally uses imports from `#imports`.

I also understand that some imports can be quite complex and are more conveniently imported from `#imports` ([composables from pinia](https://github.com/alSergey/nuxt-autoimports-repro/blob/master/.nuxt/imports.d.ts#L29) is a good example). What I’m saying is that I want to be able to disable re-exports of the `composables` and `utils` folders from my project, as well as the `vue` and `@unhead/vue` libraries. Apart from these libraries, I didn’t find anything else that could be disabled without losing convenience (all other libraries are imported through complex paths).

I looked at the `nuxt` code and found that you add these re-exports yourself, meaning this can be disabled somehow. 

This is where `composables` and `utils` are added. https://github.com/nuxt/nuxt/blob/275bee5b0812b59c20a51c73d2bc4c36d03b5185/packages/nuxt/src/imports/module.ts#L52-L63

This is where `vue` is added. https://github.com/nuxt/nuxt/blob/275bee5b0812b59c20a51c73d2bc4c36d03b5185/packages/nuxt/src/imports/presets.ts#L121-L218 

This is where `@unhead/vue` is added.
https://github.com/nuxt/nuxt/blob/275bee5b0812b59c20a51c73d2bc4c36d03b5185/packages/nuxt/src/head/module.ts#L38-L50

Considering that @danielroe [said](https://github.com/nuxt/nuxt/issues/26525#issuecomment-2025826322) that this can be done, as I understand it, the only thing that is needed on your part is to describe in the documentation how to do this.

**Comment by AndrewBogdanovTSS** [comment]
Though I'm not entirely against the opt-in nature of auto imports (with the extended possibilities of configuration) at the same time I think that this feature is one of those things they make Nuxt - Nuxt - the framework  with unparalleled DX. I represent a person who is working on a project of 100K+ lines of FE code and 65+ devs that work in more than 10 teams. By the last 1.5 years of using Nuxt 3 we had 0 issues with auto imports that we couldn't resolve but at the same time much leaner codebase, bigger opportunities due to layers and ultimate happier devs.

**Comment by antoinerey** [comment]
@AndrewBogdanovTSS That's interesting! Our team is roughly half the size of yours, and we've already hit some limitations while you did not. My understanding is that we're doing things differently from you, and it's not working as good as for you. 😅

May I ask what your project architecture and/or your processes look like? More specifically, how do you prevent two teams from overriding each other through the "global" exports mechanism available from #imports. 

I understand it's a bit off from the initial issue. But, if we could share some best practices to work with Nuxt at scale, we could even document them properly as with everything else, and ease the life of many teams out there. 

(Consider myself willing to contribute on that if that's something the Nuxt team sees as valuable)

**Comment by pi0** [comment]
Related Nitro discussion: https://github.com/unjs/nitro/issues/2232

**Comment by antoinerey** [comment]
Despite the fact that this issue is now closed (and, by the way, thanks to the Nuxt team for the `scan` option), I'm still interested in learning more about the way you structure your project @AndrewBogdanovTSS. 

If you ever come back to this issue, don't hesitate to share what seems to be working for you. 🤗

**Comment by alSergey** [comment]
@antfu hi, has the feature already been released? I looked at the latest version 3.11.2 and it's not there
